### DISCUSSION TITLE
Black screen except for the top right text on one monitor

### DISCUSSION THREAD

**Discussion by dblsaiko** [CuarzoSoftware/Louvre#45] [Tested Hardware & Platforms]
Let me finally report this :^) I originally thought this was the same as https://github.com/CuarzoSoftware/Louvre/discussions/41 when I saw that, but I don't think it is.

Louvre isn't updating the screen correctly, one monitor has a black screen except for the info text. When the clock (or I assume any of the text on the top bar) updates, which of the ones has the broken background changes, like here, these are a minute apart:
![IMG_2326](https://github.com/CuarzoSoftware/Louvre/assets/3987560/6cd0e80a-694e-4f8a-914c-d098325c416e)
![IMG_2325](https://github.com/CuarzoSoftware/Louvre/assets/3987560/61768fa0-bc1b-4de8-aa1a-c9283edc00f3)

**Comment by ehopperdietzel** [comment]
Hmm weird, it seems like the scene damage tracking is not working properly, what happens if you move the cursor to the black screen and then change the scaling factor with `Left Super + Left Shift + [Up / Down arrows]`? That should damage the entire screen, also are the displays connected to different GPUs?

**Comment by dblsaiko** [comment]
Yup, that fixes it. The screens are connected to the same GPU (RX 6800 XT).

**Comment by ehopperdietzel** [comment]
Phew, great! 😅 I am going to check `louvre-views`. I think I made some modifications in the last release and maybe forgot to do a full damage when a new display is hotplugged.

**Comment by ehopperdietzel** [comment]
I tested `louvre-views` on all my devices, using 2 displays, and I didn't see this bug. Can you tell me the context in which this happens? The situations where I would expect this to happen are:

1. Compositor Initialization
2. Display Hotplugging
3. Returning from another TTY
4. Maybe when using triple buffering (**SRM_RENDER_MODE_ITSELF_FB_COUNT**=3)

**Comment by dblsaiko** [comment]
It happens when starting the compositor. Switching TTYs or unplugging the display actually fixes it. And it seems to not happen with triple buffering.

**Comment by ehopperdietzel** [comment]
Thank you, does this happen with `louvre-weston-clone` as well? Maybe SRM is failing to do the first page flip. If you run `louvre-views` with **SRM_DEBUG**=3, are there any error messages related to faulty page flips?

**Comment by dblsaiko** [comment]
It prints nothing with SRM_DEBUG=3. louvre-weston-clone I assume has the same problem but it affects it differently: the left monitor initially shows no time, and when it updates it is always one minute behind the right screen. It shows the correct time only while you mouse over the terminal icon in the top left on that monitor.

(this is just after starting it and doing nothing else, I haven't tried unplugging any monitor)

**Comment by ehopperdietzel** [comment]
It's weird because that doesn't happen to me, so I think the damage tracking should be okay. It seems like, for some reason, some of the first OpenGL calls have no effect in your case. Each time an output is initialized, a new rendering thread is created. So, maybe I should do an initial `glFinish()` or something similar to make sure the shared OpenGL context is fully initialized before doing rendering. I'll check that out. Thank you, by the way!

**Comment by ehopperdietzel** [comment]
I've added some `glFinish()` calls after initializing the backend, after output initialization, and following the rendering of the first frame in an output. Could you check if that fixes your issue? The changes are in the `amd_gpu` branch.

**Comment by ehopperdietzel** [comment]
It could be a [driver issue](https://community.amd.com/t5/drivers-software/22-10-2-black-screens-and-driver-timeouts/m-p/552213). Perhaps AMDGPU doesn't handle multithreading properly yet.

**Comment by dblsaiko** [comment]
Looks like that patch fixes it, at least the original issue with louvre-views.

The issue with the clock in louvre-weston-clone stays the same though.

It might be a driver issue, is there a way to disable multithreaded rendering right now?

**Comment by ehopperdietzel** [comment]
Nice! Hopefully, it was just a context/thread initialization problem. Because if you use `louvre-views`, launch apps, drag them around, and such, do you see any black square glitches popping up? I am going to test `louvre-weston-clone` again. I remember seeing a similar problem with the clock, but that must be because that example does not use the scene system, so I had to implement damage tracking manually, which, of course, can lead to this kind of bug when not handled properly. Regarding your question, the answer is no, at least with the SRM backend, it is not possible to use a single thread only.

**Comment by ehopperdietzel** [comment]
I tested `louvre-weston-clone`, and the clock seems to work fine in my case. All the clocks for each display update simultaneously with the same values. Perhaps, in your case, the rendering commands may not be finishing in time when `eglSwapBuffers()` is called. However, according to [this](https://registry.khronos.org/OpenGL-Refpages/es1.1/xhtml/glFinish.xml), calling `glFinish()` is not necessary. I will try adding `glFlush()` after rendering and before calling `eglSwapBuffers()` to see if it has any effect.

**Comment by ehopperdietzel** [comment]
I pushed the change to the `amd_gpu` branch.

**Reply by dblsaiko** [reply]
Sorry it took me a while, that does fix it!

**Reply by ehopperdietzel** [reply]
Yuhuu! 🎉 Thank you once again! I think I'll add the changes to SRM instead, since it's actually an SRM bug.

**Reply by ehopperdietzel** [reply]
Don't know why GitHub didn't notify me about this

**Comment by ehopperdietzel** [comment]
Hi @2xsaiko, JFYI fences have been implemented in the latest SRM and Louvre versions, which should completely resolve this issue.
